lib/IDE/Formatting.cpp

```diff
@@ -23,85 +24,246 @@ using namespace ide;
-  }
-  operator bool() { return LineStarts.size() > 1; }
-  bool isRBraceDotsPattern() const {
-    for (auto It = LineStarts.rbegin(), end = LineStarts.rend();
-         It + 1 < end; ++It) {
-      auto* CurrentLine = *It;
-      auto* PreviousLine = *(It + 1);
-      if (CurrentLine->getKind() == tok::period &&
-          PreviousLine->getKind() == tok::period) {
-        // If the previous line starts with dot too, move further up.
-        continue;
-      } else if (CurrentLine->getKind() == tok::period &&
-                 PreviousLine->getKind() == tok::r_brace &&
-                 PreviousLine + 1 == CurrentLine) {
-        // Check if the previous line starts with '}' and the period of the
-        // current line is immediately after the '}'
-        return true;
-      } else {
-        return false;
+
+/// A helper class used to optionally override the ContextLoc and Kind of an
+/// IndentContext.
+class ContextOverride {
+  struct Override {
+    /// The overriding ContextLoc.
+    SourceLoc ContextLoc;
+    /// The overriding Kind.
+    IndentContext::ContextKind Kind;
+    /// The location after which this override takes effect.
+    SourceLoc ApplicableFrom;
+  };
+  Optional<Override> Value;
+
+public:
+  /// Sets this override to make an IndentContext indent relative to the exact
+  /// column of AlignLoc if the IndentContext's ContextLoc is >= AlignLoc and
+  /// on the same line.
+  void setExact(SourceManager &SM, SourceLoc AlignLoc) {
+    Value = {AlignLoc, IndentContext::Exact, AlignLoc};
+  }
+
+  /// Sets this override to propagate the given ContextLoc along to any
+  /// IndentContext with a ContextLoc >= L and on the same line. If this
+  /// override's existing value applies to the provided ContextLoc, its
+  /// ContextLoc is propagated instead.
+  ///
+  /// This propagation is necessary for cases like the trailing closure of 'bar'
+  /// in the example below. It's direct ContextLoc is 'bar', but we want
+  /// it to be 'foo' (the ContextLoc of its parent tuple expression):
+  ///
+  /// \code
+  /// foo(a: 1,
+  ///     b: 2)(45, bar(c: 1,
+  ///                   d: 2) {
+  ///   fatalError()
+  /// })
+  /// \endcode
+  SourceLoc propagateContext(SourceManager &SM, SourceLoc ContextLoc,
+                             IndentContext::ContextKind Kind,
+                             SourceLoc L, SourceLoc R) {
+    // If the range ends on the same line as it starts, we know up front that
+    // no child range can span multiple lines, so there's no need to propagate
+    // ContextLoc via this override.
+    if (R.isValid() && isOnSameLine(SM, L, R))
+      return ContextLoc;
+
+    Override NewValue = {ContextLoc, Kind, L};
+    if (Value) {
+      // If the existing override applies to the same line, it's ContextLoc
+      // should override this ContextLoc - don't propagate the intermediary.
+      if (isOnSameLine(SM, Value->ApplicableFrom, NewValue.ApplicableFrom)) {
```

Comment:
What does this do to your example?  E.g. if you pull bar onto the same line as foo what happens to the indentation of `d` and `fatalError`, and the closing `})`?

```
foo(a: 1, b: 2)(45, bar(c: 1,
                  d: 2) {
  fatalError()
})

Reply:
It's needed for the original example too. The parens of `bar`'s argument and the parens containing those both set overrides because they span multiple lines (the previous check). When the parens of `bar`'s argument are being processed, the outer parens have already set an override, so `Value` is set. The existing override's applicable loc is its left paren which is on the same line as `NewValue`'s applicable loc - the left paren of bar's argument. If this check wasn't here with the early return, we'd effectively end up keeping the ContextLoc and Kind we had before, but we'd change the ApplicableFrom loc to  be `bar`'s left paren.  When we move on to process `bar`'s trailing closure, which has a native `ContextLoc` of `bar`, the override doesn't apply to it, because `bar` is before the `ApplicableFrom` location of the override (`bar`'s left paren). The example you gave would still indent the same, because `bar` and `foo` are on the same line, but in the original example, the trailing closure will indent relative to the line containing 'bar', giving:
```
 foo(a: 1,
     b: 2)(45, bar(c: 1,
                   d: 2) {
       fatalError()
     })
```

I think I'll restructure this code a bit to hopefully make this clearer along with some more explanation.